# Suport Vector Machine on Breast Cancer data

* Project: Predict whether tumor is malignant or benign 
* Project Scope: Analyze,Visualize and apply Suport Vector Machine model on Breast Cancer Data
* Data source: Sklearn dataset
* The data contains the following columns:
    * 'User ID': Social Networks users login id
    * 'Gender': Gender of the user
    * 'Age': Age of the user
    * 'EstimatedSalary': Estimated Salary of the user
    * 'Purchased': User purchased (1) or not (0)

## Import libraries & Load data

In [1]:
# Import Data Analysis libraries
import pandas as pd
import numpy as np

# Import Data Visualisation libraries
import matplotlib.pyplot as plt
import seaborn as sns

# For display in notebook
%matplotlib inline

# Load data
from sklearn.datasets import load_breast_cancer
dataset = load_breast_cancer()

## Lets check the data

In [2]:
# Data is in the farm of dictionary. Lets check the keys
dataset.keys()

dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names'])

* We can see there are 5 keys
* Input variables/features (X) are in under data key and its column names are under feature_names
* Output /target variable (y) is under target key and its column names are under target_names 
* Description of the file is under DESCR

In [3]:
# Checking description of the file
print(dataset['DESCR'])

Breast Cancer Wisconsin (Diagnostic) Database

Notes
-----
Data Set Characteristics:
    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry 
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 3 is Mean Radius, field
        13 is Radius SE, field 23 is Worst Radius.

        

In [4]:
# create input and output dataframes
X = pd.DataFrame(dataset['data'],columns=dataset['feature_names'])
y = dataset['target']

In [5]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 30 columns):
mean radius                569 non-null float64
mean texture               569 non-null float64
mean perimeter             569 non-null float64
mean area                  569 non-null float64
mean smoothness            569 non-null float64
mean compactness           569 non-null float64
mean concavity             569 non-null float64
mean concave points        569 non-null float64
mean symmetry              569 non-null float64
mean fractal dimension     569 non-null float64
radius error               569 non-null float64
texture error              569 non-null float64
perimeter error            569 non-null float64
area error                 569 non-null float64
smoothness error           569 non-null float64
compactness error          569 non-null float64
concavity error            569 non-null float64
concave points error       569 non-null float64
symmetry error             569 

we can see dataset got 568 examples with 30 features

In [6]:
# lets check the data
X.head(2)

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38           122.8     1001.0          0.11840   
1        20.57         17.77           132.9     1326.0          0.08474   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   

   mean fractal dimension           ...             worst radius  \
0                 0.07871           ...                    25.38   
1                 0.05667           ...                    24.99   

   worst texture  worst perimeter  worst area  worst smoothness  \
0          17.33            184.6      2019.0            0.1622   
1          23.41            158.8      1956.0            0.1238   

   worst compactness  worst concavity  worst concave points  worst symmetry  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   

   worst fractal dimension  
0                  0.11890  
1                  0.08902  

[2 rows x 30 columns]

In [7]:
X.describe()

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension           ...             \
count     569.000000              569.000000           ...              
mean        0.181162                0.062798           ...              
std         0.027414                0.007060           ...              
min         0.106000                0.049960           ...              
25%         0.161900                0.057700           ...              
50%         0.179200                0.061540           ...              
75%         0.195700                0.066120           ...              
max         0.304000                0.097440           ...              

       worst radius  worst texture  worst perimeter   worst area  \
count    569.000000     569.000000       569.000000   569.000000   
mean      16.269190      25.677223       107.261213   880.583128   
std        4.833242       6.146258        33.602542   569.356993   
min        7.930000      12.020000        50.410000   185.200000   
25%       13.010000      21.080000        84.110000   515.300000   
50%       14.970000      25.410000        97.660000   686.500000   
75%       18.790000      29.720000       125.400000  1084.000000   
max       36.040000      49.540000       251.200000  4254.000000   

       worst smoothness  worst compactness  worst concavity  \
count        569.000000         569.000000       569.000000   
mean           0.132369           0.254265         0.272188   
std            0.022832           0.157336         0.208624   
min            0.071170           0.027290         0.000000   
25%            0.116600           0.147200         0.114500   
50%            0.131300           0.211900         0.226700   
75%            0.146000           0.339100         0.382900   
max            0.222600           1.058000         1.252000   

       worst concave points  worst symmetry  worst fractal dimension  
count            569.000000      569.000000               569.000000  
mean               0.114606        0.290076                 0.083946  
std                0.065732        0.061867                 0.018061  
min                0.000000        0.156500                 0.055040  
25%                0.064930        0.250400                 0.071460  
50%                0.099930        0.282200                 0.080040  
75%                0.161400        0.317900                 0.092080  
max                0.291000        0.663800                 0.207500  

[8 rows x 30 columns]

this gives descriptive statistics that sumarize the central tendency of numarical columns

# Implementing the model

## Splitting the dataset into the Training set and Test set

In [8]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

In [9]:
# Lets see the training and test sets
for item in [X_train, y_train, X_test, y_test]:
    print(item.shape)

(426, 30)
(426,)
(143, 30)
(143,)


## Feature Scaling

In [10]:
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.fit_transform(X_test)

In [11]:
# Lets see how the data will be after feature scaling
X_train[:2,:]

array([[-0.65079907, -0.43057322, -0.68024847, -0.62698309, -0.91381897,
        -0.93335153, -0.78763077, -0.67944444,  0.25404078, -0.79877182,
        -0.68958948,  0.04723809, -0.78036299, -0.55337361,  0.21546423,
        -0.59450332, -0.55964264, -0.52889336, -0.30718648, -0.73414209,
        -0.6661438 ,  0.01849243, -0.68117939, -0.63150142,  0.19196229,
        -0.50775064, -0.69592933, -0.36433881,  0.32349851, -0.7578486 ],
       [-0.82835341,  0.15226547, -0.82773762, -0.75309358,  0.65281216,
        -0.70875266, -1.06654609, -1.07190884,  0.48903741, -0.34145374,
        -0.37372313,  1.38156616, -0.42829675, -0.43492182,  2.09091025,
        -0.82810471, -0.90419591, -1.14032681,  2.68441903, -0.66600912,
        -0.88704517,  0.02344492, -0.90048554, -0.77100112,  0.43210093,
        -1.02916891, -1.29277423, -1.45036679,  0.62563098, -1.03071387]])

In [12]:
X_test[:2,:]

array([[-0.1839902 ,  0.22170989, -0.11761404, -0.26973359,  0.97193444,
         0.69163406,  0.66422425,  0.86379926,  1.07264367,  1.30108456,
        -0.06387943, -0.52896594,  0.08026487, -0.16010411, -0.5157891 ,
        -0.19713754,  0.07853594,  0.18366786, -0.4507456 , -0.01293877,
         0.06089154,  0.60582702,  0.20398748, -0.02939386,  1.12864697,
         0.73704174,  0.97465513,  1.40089716,  1.16977773,  1.37108056],
       [-0.23927557,  1.20953909, -0.30776593, -0.32756529, -0.57492927,
        -0.92580153, -0.73956439, -0.71148962, -0.80536136, -0.9836759 ,
        -0.80543757,  0.21044227, -0.85774844, -0.62021463, -0.41371289,
        -0.91150812, -0.66647515, -0.81890411, -0.87464651, -0.82601506,
        -0.38282965,  1.32199868, -0.47336605, -0.44009447, -0.13610472,
        -0.88286449, -0.59768168, -0.79588429, -0.81775175, -0.91991041]])

## Fitting Suport Vector Machine Model to the Training set

In [13]:
from sklearn.svm import SVC
classifier = SVC(kernel='rbf',random_state = 99)
classifier.fit(X_train, y_train)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=99, shrinking=True,
  tol=0.001, verbose=False)

In [14]:
# Lets check the accuracy of train and test results
print("Training set accuracy: {:.3f}".format(classifier.score(X_train, y_train)))
print("Test set accuracy: {:.3f}".format(classifier.score(X_test, y_test)))

Training set accuracy: 0.986
Test set accuracy: 0.965


We got good accuracy for test results (96%). But lets see if we can increace it by adding regularization and gamma parameters

## Grid search with Regularization and Gamma

The default regularization value (C)=1 and gama ='auto' gave us 82% accuracy on training dataset and 89% accuracy on test dataset. Lets try on different regularization values(C): [0.1,1, 10, 100], and Gamma values (gamma): [1,0.1,0.01,0.001]
Regularization value controls the cost of miss-classification. Large C value will penalize the cost, so will have low bias and high variance
Gama is related to 'rbf' Gaussian kernal. Large Gamma value will have high bias and low variance

In [22]:
from sklearn.model_selection import GridSearchCV
param_grid = {'C': [0.1,1, 10, 100], 'gamma': [1,0.1,0.01,0.001]}
grid = GridSearchCV(SVC(),param_grid,refit=True,verbose=2) #verbose=2 will print all the details
grid.fit(X_train,y_train)
print('Best Parameters using grid search: \n', grid.best_params_)

Fitting 3 folds for each of 16 candidates, totalling 48 fits
[CV] C=0.1, gamma=1 ..................................................
[CV] ................................... C=0.1, gamma=1, total=   0.0s
[CV] C=0.1, gamma=1 ..................................................
[CV] ................................... C=0.1, gamma=1, total=   0.0s
[CV] C=0.1, gamma=1 ..................................................
[CV] ................................... C=0.1, gamma=1, total=   0.0s
[CV] C=0.1, gamma=0.1 ................................................
[CV] ................................. C=0.1, gamma=0.1, total=   0.0s
[CV] C=0.1, gamma=0.1 ................................................
[CV] ................................. C=0.1, gamma=0.1, total=   0.0s
[CV] C=0.1, gamma=0.1 ................................................
[CV] ................................. C=0.1, gamma=0.1, total=   0.0s
[CV] C=0.1, gamma=0.01 ...............................................
[CV] ...........

[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s remaining:    0.0s


[CV] ................................... C=1, gamma=0.1, total=   0.0s
[CV] C=1, gamma=0.01 .................................................
[CV] .................................. C=1, gamma=0.01, total=   0.0s
[CV] C=1, gamma=0.01 .................................................
[CV] .................................. C=1, gamma=0.01, total=   0.0s
[CV] C=1, gamma=0.01 .................................................
[CV] .................................. C=1, gamma=0.01, total=   0.0s
[CV] C=1, gamma=0.001 ................................................
[CV] ................................. C=1, gamma=0.001, total=   0.0s
[CV] C=1, gamma=0.001 ................................................
[CV] ................................. C=1, gamma=0.001, total=   0.0s
[CV] C=1, gamma=0.001 ................................................
[CV] ................................. C=1, gamma=0.001, total=   0.0s
[CV] C=10, gamma=1 ...................................................
[CV] .

[Parallel(n_jobs=1)]: Done  48 out of  48 | elapsed:    0.3s finished


## Predicting the test results with Grid model
Now lets take this grid model to predict the test set

In [17]:
y_grid_pred = grid.predict(X_test)

## Evaluating the model

In [18]:
# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_grid_pred)
cm

array([[51,  2],
       [ 2, 88]])

In [19]:
# We can check precision,recall,f1-score using classification report
from sklearn.metrics import classification_report
print(classification_report(y_test,y_grid_pred))

             precision    recall  f1-score   support

          0       0.96      0.96      0.96        53
          1       0.98      0.98      0.98        90

avg / total       0.97      0.97      0.97       143



We can see our Support Vector Machine model accuracy has slight change after applying grid model with different regularization and gamma values. Accuray score of 97% is pretty good given the size of the dataset we have.